Add --start-in-systray: "Start application minimized in the system tray"

tuxlifan opened the pull request:
The aim of this option is to provide a way to start the Signet desktop
client e.g. during login and not have the user bothered by the window
popping up.

When needed the user shall display the window by either pressing the button
on the device, left-clicking the systray icon or chosing "Open" from the
systray icon right-click menu.

FIXME:
Currently showMinimized() is called before hide() to be able to make
the window visible when user presses the button on the device.
This introduces a flicker in the task bar since the window is for a brief
moment shown there as minimized before it is hidden.
It would be better to use setVisible() or showNormal() where the button
press is handled and just hide() in SignetApplication::init.

I could not yet figure out where the code is that handles a button press on the device to activate the main window; any pointer appreciated.

tuxlifan committed:
Add --start-in-systray: "Start application minimized in the system tray"

The aim of this option is to provide a way to start the Signet desktop
client e.g. during login and not have the user bothered by the window
popping up.
When needed the user shall display the window by either pressing the button
on the device, left-clicking the systray icon or chosing "Open" from the
systray icon right-click menu.

FIXME:
Currently showMinimized() is called before hide() to be able to make
the window visible when user presses the button on the device.
This introduces a flicker in the task bar since the window is for a brief
moment shown there as minimized before it is hidden.
It would be better to use setVisible() or showNormal() where the button
press is handled and just hide() in SignetApplication::init.

nthdimtech commented:
Regarding the flicker. Have you tried calling nothing at all? I believe the window will not be visible at all until show() is called.

As for handling the button press the call sequence is:

1) signetdev_priv_handle_device_event()
2) SignetApplication::deviceEventS()
3) SignetApplication::DeviceEvent() (signal)
4) MainWindow::signetDevEvent() (slot)
5) MainWindow::open()

You probably didn't find it because I have some magic numbers for the event numbers from the device so it wouldn't be easy to search for.

tuxlifan commented:
Thank you for the call sequence :-) I was indeed not realizing that I was blindly reading over it in mainwindow.cpp.

> Regarding the flicker. Have you tried calling nothing at all? I believe the window will not be visible at all until show() is called.

Yes, not calling anything would not show the window.
However, it would also make the button press not have any effect.
Apparently show() and showMinimized() have some side-effect?!

With some crude printf()s sprinkled around:
> $ ./signet-desktop-client/signet --start-in-systray
>   DEBUG: signetdev_linux.c:transaction_thread() called.
>   DEBUG: &nbsp; transaction_thread(): in 'while (1)'
>   DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
>   DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> 

and then it just sits there, even when the button is being pressed.

While with a show() in SignetApplication::init() I get:

> $ ./signet-desktop-client/signet
> DEBUG: signetdev_linux.c:transaction_thread() called.
> DEBUG: &nbsp; transaction_thread(): in 'while (1)'
> DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> DEBUG: &nbsp; transaction_thread(): in 'while (1)'
> DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=16.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> DEBUG: signetdev_linux.c:raw_hid_io_iter() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: &nbsp; transaction_thread(): in 'while (1)'
> DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=16.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> DEBUG: signetdev_linux.c:raw_hid_io_iter() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: &nbsp; transaction_thread(): in 'while (1)'
> DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> DEBUG: signetdev_linux.c:raw_hid_io_iter() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: signetdev_priv_process_rx_packet() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> 
> #WINDOW IS VISIBLE NOW.
> #PRESSING THE BUTTON:
> 
> DEBUG: &nbsp; transaction_thread(): in 'while (1)'
> DEBUG: signetdev_linux.c:command_pipe_io_iter() called.
> DEBUG: &nbsp; command_pipe_io_iter(): rc=-1.
> DEBUG: signetdev_linux.c:raw_hid_io_iter() called.
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> DEBUG: signetdev_priv_process_rx_packet() called.
> DEBUG: signetdev_priv_handle_device_event() called.
> debug: SignetApplication::deviceEventS() called
> DEBUG: signetdev_linux.c:raw_hid_io() called.
> DEBUG: signetdev_linux.c:attempt_raw_hid_read() called.
> debug: MainWindow::signetDevEvent() called with code= 1
> debug: MainWindow::open() called.

So it seems that with show() or showMinimized() something more is coming through the command pipe on startup which makes it unstuck?

nthdimtech commented:
> Yes, not calling anything would not show the window.
However, it would also make the button press not have any effect.
Apparently show() and showMinimized() have some side-effect?!

Yes I realize now what's happening. There was a bug a long time ago where connecting to the device before the window is displayed would cause problems. I believe it was because connection handling code makes some calls on the main window that don't behave as desired if the window has yet to be mapped by the window manager. You can see how the connection is initiated in MainWindow::showEvent(). I'll see if I can think of a workaround that safely initializes the connection without hitting the issue.

nthdimtech commented:
Looked into this some more and I'm not seeing any easy fixes. This would be an especially tricky problem on windows because the native window handle is needed to receive device connection notifications. The solution is to create a windows service handle to receive the notifications instead.

In addition to that on other OS's you would need to have independent GUI and connection states. The connection states could transition without any GUI changes. Once the window becomes visible the GUI state could transition. See MainWindow::enterDeviceState(). I don't have the time to take this on but I can keep pointing you in the right direction.

nthdimtech commented:
I've done some testing I believe it is safe on a non-windows system to move the connection code from showEvent() into the end of the MainWindow constructor. We could move forward by making this a UNIX feature only for now and once I create a service handle on the windows path the windows client can pick up the feature too. I'll want to do this anyways for when I develop a server architecture for use by Browser plugins.

tuxlifan committed:
For Q_OS_UNIX open connection already in MainWindow constructor

This shows the mainwindow on button press even without having called
show() or showMinimized() before.

tuxlifan committed:
Only parse command line on Q_OS_UNIX for now.

tuxlifan commented:
Ok, I think I have made the appropriate changes as you suggested to make it Q_OS_UNIX-only for now.

Tested on Linux with icewm and i3wm.

Rebased onto v0.9.10-rc1.

nthdimtech commented:
Looks good
